api/filrewardsd/service/service.go

```diff
@@ -0,0 +1,670 @@
+	findOpts := options.Find()
+	if req.Limit > 0 {
+		findOpts.Limit = &req.Limit
+	}
+	sort := -1
+	if req.Ascending {
+		sort = 1
+	}
+	findOpts.Sort = bson.D{primitive.E{Key: "created_at", Value: sort}}
+	filter := bson.M{}
+	if req.OrgKeyFilter != "" {
+		filter["org_key"] = req.OrgKeyFilter
+	}
+	if req.ClaimedByFilter != "" {
+		filter["claimed_by"] = req.ClaimedByFilter
+	}
+	if req.StateFilter != pb.ClaimState_CLAIM_STATE_UNSPECIFIED {
+		filter["state"] = req.StateFilter
+	}
+	comp := "$lt"
+	if req.MoreToken != 0 {
+		if req.Ascending {
+			comp = "$gt"
+		}
+		t := time.Unix(0, req.MoreToken)
+		filter["created_at"] = bson.M{comp: &t}
+	}
+	cursor, err := s.claimsCol.Find(ctx, filter, findOpts)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "querying claims: %v", err)
+	}
+	defer cursor.Close(ctx)
+	var claims []claim
+	err = cursor.All(ctx, &claims)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "decoding claim query results: %v", err)
+	}
+
+	more := false
+	var startAt *time.Time
+	if len(claims) > 0 {
+		lastCreatedAt := &claims[len(claims)-1].CreatedAt
+		filter["created_at"] = bson.M{comp: *lastCreatedAt}
+		res := s.claimsCol.FindOne(ctx, filter)
+		if res.Err() != nil && res.Err() != mongo.ErrNoDocuments {
+			return nil, status.Errorf(codes.Internal, "checking for more data: %v", err)
+		}
+		if res.Err() != mongo.ErrNoDocuments {
+			more = true
+			startAt = lastCreatedAt
+		}
+	}
+	var pbClaims []*pb.Claim
+	for _, rec := range claims {
+		pbClaims = append(pbClaims, toPbClaim(&rec))
+	}
+	res := &pb.ListClaimsResponse{
+		Claims: pbClaims,
+		More:   more,
+	}
+	if startAt != nil {
+		res.MoreToken = startAt.UnixNano()
+	}
+	return res, nil
+}
+
+func (s *Service) Balance(ctx context.Context, req *pb.BalanceRequest) (*pb.BalanceResponse, error) {
+	lck := s.semaphores.Get(orgKeyLock(req.OrgKey))
+	lck.Acquire()
+	defer lck.Release()
+
+	totalRewarded, err := s.totalRewarded(ctx, req.OrgKey)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "calculating total rewarded: %v", err)
+	}
+	totalPending, err := s.totalClaimed(ctx, req.OrgKey, pb.ClaimState_CLAIM_STATE_PENDING)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "calculating total claimed: %v", err)
+	}
+	totalClaimed, err := s.totalClaimed(ctx, req.OrgKey, pb.ClaimState_CLAIM_STATE_COMPLETE)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "calculating total claimed: %v", err)
+	}
+	return &pb.BalanceResponse{
+		Rewarded:  totalRewarded,
+		Pending:   totalPending,
+		Claimed:   totalClaimed,
+		Available: totalRewarded - totalPending - totalClaimed,
+	}, nil
+}
+
+func (s *Service) Close() {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	if err := s.rewardsCol.Database().Client().Disconnect(ctx); err != nil {
+		log.Errorf("disconnecting mongo client: %s", err)
```

Comment:
Return error?

Reply:
Same thought/quesion as in the other file where you mentioned this.